File: tools/build_quota_ledger.py
#!/usr/bin/env python3
"""Generate iqdash/data/quotaLedger.json from the IQ Dash master workbook.

WHY THIS EXISTS
---------------
quotaLedger.json drives every headline quota KPI on the dashboard (obtained,
utilization, available) via iq_apply_ledger() in iqdash_data.php. It was
hand-built on 2026-07-01 from master copy (6) and had no generator, so it
silently froze while the master moved on: by 2026-07-27 the dashboard read
33,730 / 18,346 against a master at 34,240 / 22,547.

MODEL
-----
The master's `Status Submisson` sheet is a block per company. Row 2 carries
product names, row 3 the HS codes; each block's rows are labelled in column C:

    Submit #N            applied-for volume        (ignored - not granted)
    Obtained #N          granted volume            -> obtained
    Revision #N          +/- moves between products -> obtained (signed)
    Utilization (MT)     allocated to customers    -> util
    Available (MT)       obtained - util           (derived, ignored)

`_meta.view` is "effective (obtained incl. revisions)", so Revision rows are
summed into obtained. A company block starts where column A holds a number and
column B a name; the code is the leading token, so "AMP (SUJU)" -> AMP.

USAGE
    python tools/build_quota_ledger.py <master.xlsx> [-o out.json] [--check ref.json]

--check compares against an existing ledger instead of writing: use it to prove
the generator reproduces a known-good file before trusting it on new input.
"""
import argparse
import json
import os
import re
import sys
from collections import OrderedDict

try:
    import openpyxl
except ImportError:
    sys.exit("needs openpyxl:  pip install openpyxl")

NAME_ROW, HS_ROW = 2, 3
COL_NO, COL_COMPANY, COL_LABEL = 1, 2, 3
TOTAL_HDR = "JUMLAH (MT)"


def clean(v):
    return str(v).strip() if v is not None else ""


def num(v):
    if isinstance(v, (int, float)):
        return float(v)
    s = clean(v).replace(",", "")
    if not s or s.upper() in ("TBA", "-"):
        return 0.0
    try:
        return float(s)
    except ValueError:
        return 0.0


def build(path):
    wb = openpyxl.load_workbook(path, data_only=True)
    ws = wb["Status Submisson"]

    # HS-coded product columns: every column whose row-3 cell looks like an HS code
    cols = OrderedDict()          # col -> hs
    products = OrderedDict()      # hs  -> product name
    for c in range(1, ws.max_column + 1):
        hs = clean(ws.cell(row=HS_ROW, column=c).value)
        if not re.match(r"^\d{4}[.]?\d{2}[.]?\d{2}$|^\d{8,10}$", hs.replace(" ", "")):
            continue
        name = clean(ws.cell(row=NAME_ROW, column=c).value)
        if clean(name) == TOTAL_HDR:
            continue
        name = re.sub(r"\s+", " ", name).strip()
        cols[c] = hs
        products.setdefault(hs, name)

    companies = OrderedDict()
    cur = None
    for r in range(1, ws.max_row + 1):
        a = ws.cell(row=r, column=COL_NO).value
        b = clean(ws.cell(row=r, column=COL_COMPANY).value)
        if isinstance(a, (int, float)) and b:
            tok = re.split(r"[\s(]+", b)[0].upper()
            if re.match(r"^[A-Z]{2,6}$", tok):
                cur = tok
                companies.setdefault(cur, OrderedDict())
        if not cur:
            continue
        lab = clean(ws.cell(row=r, column=COL_LABEL).value)
        if re.match(r"^Obtained", lab, re.I) or re.match(r"^Revision", lab, re.I):
            field = "obtained"
        elif re.match(r"^Utili", lab, re.I) and "(MT)" in lab:
            field = "util"
        else:
            continue
        for c, hs in cols.items():
            v = num(ws.cell(row=r, column=c).value)
            if v == 0:
                continue
            ent = companies[cur].setdefault(hs, {"obtained": 0.0, "util": 0.0})
            ent[field] += v
    wb.close()

    # drop empties, normalise -0.0 and integral floats
    out = OrderedDict()
    for co, ent in companies.items():
        keep = OrderedDict()
        for hs, v in ent.items():
            o, u = round(v["obtained"], 3) + 0.0, round(v["util"], 3) + 0.0
            if o == 0 and u == 0:
                continue
            keep[hs] = {"obtained": int(o) if o == int(o) else o,
                        "util": int(u) if u == int(u) else u}
        if keep:
            out[co] = keep
    used = {hs for ent in out.values() for hs in ent}
    return {"products": OrderedDict((k, v) for k, v in products.items() if k in used),
            "companies": out}


def totals(led):
    o = sum(v["obtained"] for ent in led["companies"].values() for v in ent.values())
    u = sum(v["util"] for ent in led["companies"].values() for v in ent.values())
    return o, u


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("master")
    ap.add_argument("-o", "--out")
    ap.add_argument("--check", help="compare against this ledger instead of writing")
    a = ap.parse_args()

    led = build(a.master)
    o, u = totals(led)
    print(f"parsed {len(led['companies'])} companies, {len(led['products'])} products")
    print(f"  obtained {o:,.0f}   util {u:,.0f}   available {o - u:,.0f}")

    if a.check:
        ref = json.load(open(a.check, encoding="utf-8"))
        ro, ru = totals(ref)
        print(f"reference: obtained {ro:,.0f}   util {ru:,.0f}")
        diffs = []
        for co in sorted(set(led["companies"]) | set(ref["companies"])):
            g, e = led["companies"].get(co, {}), ref["companies"].get(co, {})
            for hs in sorted(set(g) | set(e)):
                gv, ev = g.get(hs, {"obtained": 0, "util": 0}), e.get(hs, {"obtained": 0, "util": 0})
                if gv["obtained"] != ev["obtained"] or gv["util"] != ev["util"]:
                    diffs.append(f"  {co:5} {hs:12} generated {gv} != reference {ev}")
        print(f"\nper-(company,HS) mismatches: {len(diffs)}")
        for d in diffs[:40]:
            print(d)
        return 0 if not diffs else 1

    led_out = OrderedDict()
    led_out["_meta"] = {
        "source": "master " + os.path.basename(a.master),
        "generated": os.environ.get("LEDGER_DATE", ""),
        "view": "effective (obtained incl. revisions)",
        "generator": "tools/build_quota_ledger.py",
    }
    led_out["products"] = led["products"]
    led_out["companies"] = led["companies"]
    dst = a.out or "iqdash/data/quotaLedger.json"
    with open(dst, "w", encoding="utf-8") as fh:
        json.dump(led_out, fh, indent=1, ensure_ascii=False)
        fh.write("\n")
    print(f"wrote {dst}")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: tools/build_quota_ledger.js
#!/usr/bin/env node
/**
 * Node port of tools/build_quota_ledger.py — generates iqdash/data/quotaLedger.json
 * from the IQ Dash master workbook. Same model, same output shape; see the
 * Python file's docstring for WHY the ledger exists and how the master is laid
 * out. This port exists because the machine that owns the master has Node and
 * the vendored SheetJS but no Python/openpyxl; keep the two in step, or delete
 * whichever one stops being used.
 *
 *   node tools/build_quota_ledger.js <master.xlsx> [-o out.json]
 *        [--check ref.json] [--names-from ref.json]
 *
 * --check compares against an existing ledger instead of writing. Run it
 * against the current quotaLedger.json before trusting a regenerate: every
 * mismatch it prints must be one you can explain from a master edit.
 */
'use strict';
const fs = require('fs');
const path = require('path');
const XLSX = require(path.join(__dirname, '..', 'iqdash', 'assets', 'vendor', 'xlsx.full.min.js'));

const NAME_ROW = 1, HS_ROW = 2;              // 0-indexed (Python used 1-indexed 2 / 3)
const COL_NO = 0, COL_COMPANY = 1, COL_LABEL = 2;
const TOTAL_HDR = 'JUMLAH (MT)';

const clean = v => (v === null || v === undefined) ? '' : String(v).trim();

function num(v) {
  if (typeof v === 'number') return v;
  const s = clean(v).replace(/,/g, '');
  if (!s || ['TBA', '-'].includes(s.toUpperCase())) return 0;
  const n = parseFloat(s);
  return isNaN(n) ? 0 : n;
}

function build(file) {
  const wb = XLSX.read(fs.readFileSync(file), { type: 'buffer' });
  const rows = XLSX.utils.sheet_to_json(wb.Sheets['Status Submisson'], { header: 1, raw: true, defval: null, blankrows: true });
  const width = rows.reduce((m, r) => Math.max(m, (r || []).length), 0);

  const cols = new Map();        // col index -> hs
  const products = new Map();    // hs -> product name
  for (let c = 0; c < width; c++) {
    const hs = clean((rows[HS_ROW] || [])[c]);
    if (!/^\d{4}[.]?\d{2}[.]?\d{2}$|^\d{8,10}$/.test(hs.replace(/ /g, ''))) continue;
    let name = clean((rows[NAME_ROW] || [])[c]);
    if (name === TOTAL_HDR) continue;
    name = name.replace(/\s+/g, ' ').trim();
    cols.set(c, hs);
    if (!products.has(hs)) products.set(hs, name);
  }

  const companies = new Map();
  let cur = null;
  for (let r = 0; r < rows.length; r++) {
    const row = rows[r] || [];
    const a = row[COL_NO], b = clean(row[COL_COMPANY]);
    if (typeof a === 'number' && b) {
      const tok = b.split(/[\s(]+/)[0].toUpperCase();
      if (/^[A-Z]{2,6}$/.test(tok)) {
        cur = tok;
        if (!companies.has(cur)) companies.set(cur, new Map());
      }
    }
    if (!cur) continue;
    const lab = clean(row[COL_LABEL]);
    let field = null;
    if (/^Obtained/i.test(lab) || /^Revision/i.test(lab)) field = 'obtained';
    else if (/^Utili/i.test(lab) && lab.includes('(MT)')) field = 'util';
    else continue;
    for (const [c, hs] of cols) {
      const v = num(row[c]);
      if (v === 0) continue;
      const ent = companies.get(cur);
      if (!ent.has(hs)) ent.set(hs, { obtained: 0, util: 0 });
      ent.get(hs)[field] += v;
    }
  }

  const r3 = x => Math.round(x * 1000) / 1000 + 0;
  const out = new Map();
  for (const [co, ent] of companies) {
    const keep = new Map();
    for (const [hs, v] of ent) {
      const o = r3(v.obtained), u = r3(v.util);
      if (o === 0 && u === 0) continue;
      keep.set(hs, { obtained: o, util: u });
    }
    if (keep.size) out.set(co, keep);
  }
  const used = new Set();
  for (const ent of out.values()) for (const hs of ent.keys()) used.add(hs);
  const prodOut = new Map();
  for (const [hs, name] of products) if (used.has(hs)) prodOut.set(hs, name);
  return { products: prodOut, companies: out };
}

const totals = led => {
  let o = 0, u = 0;
  for (const ent of led.companies.values()) for (const v of ent.values()) { o += v.obtained; u += v.util; }
  return [o, u];
};
const toObj = led => ({
  products: Object.fromEntries(led.products),
  companies: Object.fromEntries([...led.companies].map(([k, v]) => [k, Object.fromEntries(v)])),
});

// ── CLI ─────────────────────────────────────────────────────────────────────
const argv = process.argv.slice(2);
const flag = n => { const i = argv.indexOf(n); return i >= 0 ? argv[i + 1] : null; };
const master = argv.find(a => !a.startsWith('-') && argv[argv.indexOf(a) - 1] !== '-o'
  && argv[argv.indexOf(a) - 1] !== '--check' && argv[argv.indexOf(a) - 1] !== '--names-from');
if (!master) { console.error('usage: node tools/build_quota_ledger.js <master.xlsx> [-o out] [--check ref] [--names-from ref]'); process.exit(2); }

const led = build(master);
const namesFrom = flag('--names-from');
if (namesFrom && fs.existsSync(namesFrom)) {
  const prev = JSON.parse(fs.readFileSync(namesFrom, 'utf8')).products || {};
  const kept = [...led.products.keys()].filter(hs => hs in prev);
  kept.forEach(hs => led.products.set(hs, prev[hs]));
  const nw = [...led.products.keys()].filter(hs => !(hs in prev));
  console.log(`names: kept ${kept.length} curated, ${nw.length} new from master`
    + (nw.length ? ` -> ${JSON.stringify(nw.map(h => led.products.get(h)))}` : ''));
}
const [o, u] = totals(led);
console.log(`parsed ${led.companies.size} companies, ${led.products.size} products`);
console.log(`  obtained ${o.toLocaleString('en-US')}   util ${u.toLocaleString('en-US')}   available ${(o - u).toLocaleString('en-US')}`);

const check = flag('--check');
if (check) {
  const ref = JSON.parse(fs.readFileSync(check, 'utf8'));
  const ro = Object.values(ref.companies).reduce((s, e) => s + Object.values(e).reduce((t, v) => t + v.obtained, 0), 0);
  const ru = Object.values(ref.companies).reduce((s, e) => s + Object.values(e).reduce((t, v) => t + v.util, 0), 0);
  console.log(`reference: obtained ${ro.toLocaleString('en-US')}   util ${ru.toLocaleString('en-US')}`);
  const g = toObj(led).companies, e = ref.companies, diffs = [];
  for (const co of [...new Set([...Object.keys(g), ...Object.keys(e)])].sort()) {
    const G = g[co] || {}, E = e[co] || {};
    for (const hs of [...new Set([...Object.keys(G), ...Object.keys(E)])].sort()) {
      const gv = G[hs] || { obtained: 0, util: 0 }, ev = E[hs] || { obtained: 0, util: 0 };
      if (gv.obtained !== ev.obtained || gv.util !== ev.util)
        diffs.push(`  ${co.padEnd(5)} ${hs.padEnd(12)} generated ${JSON.stringify(gv)} != reference ${JSON.stringify(ev)}`);
    }
  }
  console.log(`\nper-(company,HS) mismatches: ${diffs.length}`);
  diffs.slice(0, 40).forEach(d => console.log(d));
  process.exit(diffs.length ? 1 : 0);
}

const outObj = {
  _meta: {
    source: 'master ' + path.basename(master),
    generated: process.env.LEDGER_DATE || '',
    view: 'effective (obtained incl. revisions)',
    generator: 'tools/build_quota_ledger.js',
  },
  ...toObj(led),
};
const dst = flag('-o') || 'iqdash/data/quotaLedger.json';
fs.writeFileSync(dst, JSON.stringify(outObj, null, 1) + '\n', 'utf8');
console.log('wrote ' + dst);
